// ngx-carousel-lib/esm5/lib/directives/dom-change.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, Output } from '@angular/core';
var DomChangeDirective = /** @class */ (function () {
    function DomChangeDirective(elementRef) {
        var _this = this;
        this.elementRef = elementRef;
        this.domChange = new EventEmitter();
        /** @type {?} */
        var element = this.elementRef.nativeElement;
        this.changes = new MutationObserver((/**
         * @param {?} mutations
         * @return {?}
         */
        function (mutations) {
            mutations.forEach((/**
             * @param {?} mutation
             * @return {?}
             */
            function (mutation) { return _this.domChange.emit(mutation); }));
        }));
        this.changes.observe(element, {
            attributes: true,
            childList: true,
            characterData: true
        });
    }
    /**
     * @return {?}
     */
    DomChangeDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.changes.disconnect();
    };
    DomChangeDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[domChange]'
                },] }
    ];
    /** @nocollapse */
    DomChangeDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    DomChangeDirective.propDecorators = {
        domChange: [{ type: Output }]
    };
    return DomChangeDirective;
}());
export { DomChangeDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DomChangeDirective.prototype.changes;
    /** @type {?} */
    DomChangeDirective.prototype.domChange;
    /**
     * @type {?}
     * @private
     */
    DomChangeDirective.prototype.elementRef;
}
