// ngx-carousel-lib/esm2015/lib/model/Carousel.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * :tmtfactory) © 2017
 * Alex Marcos <alejandro.marcos@tmtfactory.com>
 * @ignore
 */
export default class Carousel {
    constructor() {
        this._currdeg = 0;
        this._totalItems = 0;
        this._maxWidthSize = 0;
        this._maxHeightSize = 0;
        this._maxDegree = 0;
        this._totalWidth = 0;
        this._isHorizontal = false;
        this._items = [];
        this._degreesSlides = [];
        this._activeIndex = 0;
        this._lastIndex = -1;
        this._lockSlides = false;
        this._autoPlayIsRunning = false;
    }
    /**
     * @return {?}
     */
    get autoPlayIsRunning() {
        return this._autoPlayIsRunning;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set autoPlayIsRunning(value) {
        this._autoPlayIsRunning = value;
    }
    /**
     * @return {?}
     */
    get currdeg() {
        return this._currdeg;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set currdeg(value) {
        this._currdeg = value;
    }
    /**
     * @return {?}
     */
    get totalItems() {
        return this._totalItems;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set totalItems(value) {
        this._totalItems = value;
    }
    /**
     * @return {?}
     */
    get isHorizontal() {
        return this._isHorizontal;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set isHorizontal(value) {
        this._isHorizontal = value;
    }
    /**
     * @return {?}
     */
    get maxWidthSize() {
        return this._maxWidthSize;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set maxWidthSize(value) {
        this._maxWidthSize = value;
    }
    /**
     * @return {?}
     */
    get maxHeightSize() {
        return this._maxHeightSize;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set maxHeightSize(value) {
        this._maxHeightSize = value;
    }
    /**
     * @return {?}
     */
    get maxDegree() {
        return this._maxDegree;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set maxDegree(value) {
        this._maxDegree = value;
    }
    /**
     * @return {?}
     */
    get totalWidth() {
        return this._totalWidth;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set totalWidth(value) {
        this._totalWidth = value;
    }
    /**
     * @return {?}
     */
    get items() {
        return this._items;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set items(value) {
        this._items = value;
    }
    /**
     * @return {?}
     */
    get degreesSlides() {
        return this._degreesSlides;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set degreesSlides(value) {
        this._degreesSlides = value;
    }
    /**
     * @return {?}
     */
    get activeIndex() {
        return this._activeIndex;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set activeIndex(value) {
        this._activeIndex = value;
    }
    /**
     * @return {?}
     */
    get lockSlides() {
        return this._lockSlides;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set lockSlides(value) {
        this._lockSlides = value;
    }
    /**
     * @return {?}
     */
    get lastIndex() {
        return this._lastIndex;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set lastIndex(value) {
        this._lastIndex = value;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    Carousel.prototype._currdeg;
    /**
     * @type {?}
     * @private
     */
    Carousel.prototype._totalItems;
    /**
     * @type {?}
     * @private
     */
    Carousel.prototype._maxWidthSize;
    /**
     * @type {?}
     * @private
     */
    Carousel.prototype._maxHeightSize;
    /**
     * @type {?}
     * @private
     */
    Carousel.prototype._maxDegree;
    /**
     * @type {?}
     * @private
     */
    Carousel.prototype._totalWidth;
    /**
     * @type {?}
     * @private
     */
    Carousel.prototype._isHorizontal;
    /**
     * @type {?}
     * @private
     */
    Carousel.prototype._items;
    /**
     * @type {?}
     * @private
     */
    Carousel.prototype._degreesSlides;
    /**
     * @type {?}
     * @private
     */
    Carousel.prototype._activeIndex;
    /**
     * @type {?}
     * @private
     */
    Carousel.prototype._lastIndex;
    /**
     * @type {?}
     * @private
     */
    Carousel.prototype._lockSlides;
    /**
     * @type {?}
     * @private
     */
    Carousel.prototype._autoPlayIsRunning;
}
