// ngx-carousel-lib/esm2015/lib/directives/dom-change.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, Output } from '@angular/core';
export class DomChangeDirective {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.domChange = new EventEmitter();
        /** @type {?} */
        const element = this.elementRef.nativeElement;
        this.changes = new MutationObserver((/**
         * @param {?} mutations
         * @return {?}
         */
        (mutations) => {
            mutations.forEach((/**
             * @param {?} mutation
             * @return {?}
             */
            (mutation) => this.domChange.emit(mutation)));
        }));
        this.changes.observe(element, {
            attributes: true,
            childList: true,
            characterData: true
        });
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.changes.disconnect();
    }
}
DomChangeDirective.decorators = [
    { type: Directive, args: [{
                selector: '[domChange]'
            },] }
];
/** @nocollapse */
DomChangeDirective.ctorParameters = () => [
    { type: ElementRef }
];
DomChangeDirective.propDecorators = {
    domChange: [{ type: Output }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    DomChangeDirective.prototype.changes;
    /** @type {?} */
    DomChangeDirective.prototype.domChange;
    /**
     * @type {?}
     * @private
     */
    DomChangeDirective.prototype.elementRef;
}
